photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/OWLQN.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.optimization
+
+import breeze.linalg.Vector
+import breeze.optimize.{OWLQN => BreezeOWLQN}
+import org.apache.spark.broadcast.Broadcast
+
+import com.linkedin.photon.ml.normalization.NormalizationContext
+
+/**
+ * L1 regularization has to be done differently because the gradient of L1 penalty term may be discontinuous.
+ * For optimization with L1 penalty term, the optimization algorithm is a modified Quasi-Newton algorithm called OWL-QN.
+ * Reference: [[http://research.microsoft.com/en-us/downloads/b1eb1016-1738-4bd5-83a9-370c9d498a03/]]
+ *
+ * @param l1RegWeight The L1 regularization weight
+ * @param normalizationContext The normalization context
+ * @param numCorrections The number of corrections used in the LBFGS update. Default 10. Values of numCorrections less
+ *                       than 3 are not recommended; large values of numCorrections will result in excessive computing
+ *                       time.
+ *                       Recommended:  3 < numCorrections < 10
+ *                       Restriction:  numCorrections > 0
+ * @param tolerance The tolerance threshold for improvement between iterations as a percentage of the initial loss
+ * @param maxNumIterations The cut-off for number of optimization iterations to perform.
+ * @param constraintMap (Optional) The map of constraints on the feature coefficients
+ * @param isTrackingState Whether to track intermediate states during optimization
+ * @param isReusingPreviousInitialState Whether to reuse the previous initial state or not. When warm-start training is
+ *                                      desired, i.e. in grid-search based hyper-parameter tuning, this field is
+ *                                      recommended to set to true for consistent convergence check.
+ */
+class OWLQN(
```

Comment:
OWLQN is its own optimizer now. The OptimizerFactory determines which one to use at runtime based on the RegularizationContext provided by the user.